### PR TITLE
Out-of-sample prediction using pseudocount

### PR DESCRIPTION
**Pull request by hylkedonker**
This pull request should resolve [#835]. 
In detail:
- `pseudocount` support for the `BernoulliDistribution` was implemented.
- The parameter `alpha` was introduced to smooth only the individual distributions but not the label (like in scikit-learn), leaving the current behaviour of `pseudocount` in tact.
- Supporting tests were written.
- No failing tests when running `python3 setup.py test`.
- scikit-learn and pomegranate now give identical predictions:
```Python
from numpy import array, testing
from pomegranate import *
from sklearn.naive_bayes import BernoulliNB

X_train = array([
    [1., 0., 1., 0., 1.],
    [1., 0., 1., 0., 1.],
    [1., 0., 1., 0., 1.],
])
y_train = array([0, 0, 1])
X_test = array([[1., 0., 1., 0., 0.]])

pmodel = NaiveBayes.from_samples(
    BernoulliDistribution, 
    X_train, 
    y_train,
    alpha=1.0,
)
skmodel = BernoulliNB(alpha=1.0).fit(X_train, y_train)

testing.assert_array_almost_equal(
    skmodel.predict_proba(X_test),  
    pmodel.predict_proba(X_test),
)
```

### CODE REVIEW

File: pomegranate/NaiveBayes.pyx [hunk 1]
```diff
@@ -93,8 +95,9 @@ cdef class NaiveBayes(BayesModel):
 
 	@classmethod
 	def from_samples(cls, distributions, X, y=None, weights=None,
-		pseudocount=0.0, stop_threshold=0.1, max_iterations=1e8,
-		callbacks=[], return_history=False, verbose=False, n_jobs=1):
+		pseudocount=0.0, alpha: Optional[float] = None, stop_threshold=0.1,
```
Comment (jmschrei) [inline pomegranate/NaiveBayes.pyx:98]: Would you mind removing the typing? Regardless of whether typing is good or bad (it's own debate), partially typing pomegranate seems like a bad idea and I can't commit to typing every parameter.

File: pomegranate/NaiveBayes.pyx [hunk 2]
```diff
@@ -140,6 +143,10 @@ cdef class NaiveBayes(BayesModel):
 			if they don't happen to occur in the data. Only effects mixture
 			models defined over discrete distributions. Default is 0.
 
+        alpha : double, optional
```
Comment (jmschrei) [inline pomegranate/NaiveBayes.pyx:146]: Is this indentation inconsistent with the previous lines?

File: pomegranate/bayes.pyx [hunk 3]
```diff
@@ -550,8 +552,8 @@ cdef class BayesModel(Model):
         free(r)
 
     def fit(self, X, y=None, weights=None, inertia=0.0, pseudocount=0.0,
-        stop_threshold=0.1, max_iterations=1e8, callbacks=[],
-        return_history=False, verbose=False, n_jobs=1):
+        alpha: Optional[float] = None, stop_threshold=0.1, max_iterations=1e8,
```
Comment (jmschrei) [inline pomegranate/bayes.pyx:555]: Same here

File: pomegranate/bayes.pyx [hunk 4]
```diff
@@ -769,7 +775,13 @@ cdef class BayesModel(Model):
         int column_idx, int d) nogil:
         return -1
 
-    def from_summaries(self, inertia=0.0, pseudocount=0.0, **kwargs):
+    def from_summaries(
+        self,
```
Comment (jmschrei) [inline pomegranate/bayes.pyx:779]: I don't think every parameter needs its own line here.

File: pomegranate/bayes.pyx [hunk 5]
```diff
@@ -769,7 +775,13 @@ cdef class BayesModel(Model):
...
+        inertia=0.0,
+        pseudocount=0.0,
+        alpha: Optional[float] = None,
```
Comment (jmschrei) [inline pomegranate/bayes.pyx:782]: Please remove typing

File: pomegranate/bayes.pyx [hunk 6]
```diff
@@ -804,8 +820,14 @@ cdef class BayesModel(Model):
...
 
         for i, distribution in enumerate(self.distributions):
-            if isinstance(distribution, DiscreteDistribution):
-                distribution.from_summaries(inertia, pseudocount)
+            if isinstance(
+                distribution,
+                (DiscreteDistribution, BernoulliDistribution, IndependentComponentsDistribution),
+            ):
+                distribution_smoothing = pseudocount
+                if alpha is not None:
```
Comment (jmschrei) [inline pomegranate/bayes.pyx:828]: I'm not sure I like this. You're saying the parameter "pseudocount" and "alpha" will do the same thing here? I'd suggest having pseudocount consistent refer to the labels and alpha refer to the distributions alone, as you mentioned in the documentation.

File: pomegranate/distributions/BernoulliDistribution.pyx [hunk 7]
```diff
@@ -75,18 +75,40 @@ cdef class BernoulliDistribution(Distribution):
...
 
-	def from_summaries(self, inertia=0.0):
+	def from_summaries(self, inertia=0.0, pseudocount: float = 0.0):
 		"""Update the parameters of the distribution from the summaries."""
 
 		if self.summaries[0] < 1e-8 or self.frozen:
 			return
 
-		p = self.summaries[1] / self.summaries[0]
+		n: int = 2
```
Comment (jmschrei) [inline pomegranate/distributions/BernoulliDistribution.pyx:84]: please remove typing

File: pomegranate/distributions/BernoulliDistribution.pyx [hunk 8]
```diff
@@ -75,18 +75,40 @@ cdef class BernoulliDistribution(Distribution):
...
+		p = (self.summaries[1] + pseudocount) / (self.summaries[0] + n * pseudocount)
 		self.p = self.p * inertia + p * (1-inertia)
 		self.logp[0] = _log(1-p)
 		self.logp[1] = _log(p)
 		self.summaries = [0.0, 0.0]
 
+	def fit(
```
Comment (jmschrei) [inline pomegranate/distributions/BernoulliDistribution.pyx:91]: Not sure a new fit function is needed here. The other classes call the `from_summaries` method not the `fit` method regardless.

File: pomegranate/distributions/IndependentComponentsDistribution.pyx [hunk 9]
```diff
@@ -243,8 +245,14 @@ cdef class IndependentComponentsDistribution(MultivariateDistribution):
 		}
 
 	@classmethod
-	def from_samples(cls, X, weights=None, distribution_weights=None,
-		pseudocount=0.0, distributions=None):
+	def from_samples(
```
Comment (jmschrei) [inline pomegranate/distributions/IndependentComponentsDistribution.pyx:248]: see previous comments

File: pomegranate/distributions/IndependentComponentsDistribution.pyx [hunk 10]
```diff
@@ -253,10 +261,24 @@ cdef class IndependentComponentsDistribution(MultivariateDistribution):
 		X, weights = weight_set(X, weights)
 		n, d = X.shape
 
-		if callable(distributions):
-			distributions = [distributions.from_samples(X[:,i], weights) for i in range(d)]
-		else:
-			distributions = [distributions[i].from_samples(X[:,i], weights) for i in range(d)]
+		initialised_distributions = []
```
Comment (jmschrei) [inline pomegranate/distributions/IndependentComponentsDistribution.pyx:264]: I'm not sure I like all the additions. Maybe just change it to something like `distributions = [distributions.from_samples(X[:,i], weights, **params) for i in range(d)]` where params is {} normally or {'pseudocount': pseudocount} if that's passed in?

File: tests/test_distributions.py [hunk 11]
```diff
@@ -879,6 +879,19 @@ def test_bernoulli():
...
 
+
+@with_setup(setup, teardown)
+def test_bernoulli_pseudocount():
+	"""
+	Test fitting Bernoulli distribution with non-zero pseudocount.
+	"""
+	a = [0.0, 0.0, 0.0]
+	d = BernoulliDistribution.from_samples(a, pseudocount=1)
+	assert_equal(d.probability(0), 0.8)
```
Comment (jmschrei) [inline tests/test_distributions.py:890]: Should probably be `assert_almost_equal` to handle floating point issues

File: tests/test_naive_bayes.py [hunk 12]
```diff
@@ -810,4 +810,89 @@ def test_io_from_samples():
...
+	# Test that both `alpha` and `pseudocount` are propagated to the
+	# BernoulliDistribution.
+	for kwargs in ({'alpha': 1.0}, {'pseudocount': 1.0}):
+		pmodel = NaiveBayes.from_samples(
+			BernoulliDistribution,
+			X_train,
+			y_train,
+			**kwargs
+		)
+		# Check fit for y=0 label (2 records + pseudocount=1 per category).
```
Comment (jmschrei) [inline tests/test_naive_bayes.py:835]: This test will probably need to be changed. We shouldn't have two parameters that do the same thing. pseudocount should be on the labels (and so only affect the NaiveBayes priors) and alpha should be on the counts for the underlying distributions. That way pseudocount works the exact same way for normal distributions as bernoulli.

**Review by jmschrei — COMMENTED** [review]
Thanks for the contribution! It looks good, in general. As I mention in the comments, would you mind (1) removing typing, (2) making sure the indentation is correct, and slightly reworking the psuedocount parameter? Let me know if you have any questions!

**Comment by hylkedonker** [comment]
Thanks for your detailed review. I'll get back to you soon.
Hylke

File: pomegranate/distributions/BernoulliDistribution.pyx (hunk 8 above)
Comment (hylkedonker) [reply pomegranate/distributions/BernoulliDistribution.pyx:91]: The `Distribution.from_samples` calls fit, so it is helpful for doing
```Python
a = [0.0, 0.0, 0.0]
d = BernoulliDistribution.from_samples(a, pseudocount=1)
```
like in the unit test.

**Comment by hylkedonker** [comment]
Hi Jacob,

I have made all the requested changes, plus I updated the `BayesClassifier` model, which was missing from the previous commit.
I hope this looks okay for you.

Kind regards,

Hylke

**Comment by hylkedonker** [comment]
Can the failing CI be related to the new release of joblib (it looks like https://github.com/scikit-learn-contrib/hdbscan/issues/436)?

**Comment by jmschrei** [comment]
Thank you for your contribution. However, pomegranate has recently been rewritten from the ground up to use PyTorch instead of Cython, and so this PR is unfortunately out of date and will be closed.
